fix AGE error when resource has no creationTimeStamp

hurf opened the pull request:
Fixes: #12581

If the resource doesn't contain creationTimeStamp, report AGE as
`unknown`. This usually happens when the resouces is created locally
or is not accually created at server side.

hurf committed:
fix AGE error when resource has no creationTimeStamp

If the resource doesn't contain creationTimeStamp, report AGE as
unknown. This usually happens when the resouces is created locally
or is not accually created at server side.

hurf commented:
@bgrant0607 - I didn't add creationTimeStamp for local object, considering they may not accually created in server side, eg. with `--dry-run`.

wojtek-t commented:
@k8s-bot ok to test

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -376,6 +376,9 @@ func shortHumanDuration(d time.Duration) string {
 // human-readable approximation.
 func translateTimestamp(timestamp util.Time) string {
+	if timestamp.IsZero() {
+		return "unknown"
```
Please use `<unknown>`, to be consistent with places where we use `<none>`.

bgrant0607 commented on pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -264,10 +264,9 @@ func ExamplePrintReplicationControllerWithNamespace() {
-			CreationTimestamp: util.Time{Time: time.Now().AddDate(-10, 0, 0)},
+			Name:      "foo",
+			Namespace: "beep",
+			Labels:    map[string]string{"foo": "bar"},
```
Please create a new test case rather than changing the existing test case.

bgrant0607 commented:
Thanks for the fix, @hurf.

hurf commented:
When writing the test, I noticed that [shortHumanDuration](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L363) can not handle times after current time, it will return something like `-65521s`, should I also add this support to the function?

bgrant0607 commented:
That would be great, @hurf. Within some epsilon (1s? 2s?), the time should probably be counted as "now", so the age would be 0s. Greater a difference than that is probably a bug, and we should print `<invalid>` or something similar. The user could then do `-o json` if they wanted to get the exact value to debug the problem.

hurf committed:
Enhance shortHumanDuration to handle more situation

Allow 2 seconds deviation of current time. For time more than 2
seconds from current time, output as '<invalid>'.

JanetKuo commented:
Looks good. Shippable appears to be a flake. Restarting it.

hurf commented:
@JanetKuo - If it really looks good to you, can you put a lgtm label? :-)

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -361,7 +361,11 @@ func podHostString(host, ip string) string {
 
 func shortHumanDuration(d time.Duration) string {
-	if seconds := int(d.Seconds()); seconds < 60 {
+	if seconds := int(d.Seconds()); seconds < -2 {
```
Please add a comment about why -2.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -361,7 +361,11 @@ func podHostString(host, ip string) string {
-	if seconds := int(d.Seconds()); seconds < 60 {
+	if seconds := int(d.Seconds()); seconds < -2 {
+		return fmt.Sprintf("<invalid>")
+	} else if seconds < 0 {
```
<=

bgrant0607 commented:
Also, please squash the commits.

bgrant0607 commented:
cc @kubernetes/kubectl

hurf commented:
With `<-2` means no more than 3 seconds(3 seconds not included), just try to keep same style with other clauses, but it's fine to reduce it to 2 seconds.

hurf commented:
Patch updated and rebased. I didn't squash the commits because they're logically separated.
